File: scripts/create_reference_samples.py
import json
import wave
from pathlib import Path

PROFILE_PATH = Path("auto_speech_data/profiles/default/profile.json")
REFERENCE_DIR = Path("auto_speech_data/profiles/default/reference")
REFERENCE_PROMPT = "这是我的声音样本，用来帮助软件生成自然清晰的中文旁白。"


def trim_wav(input_path: Path, output_path: Path, seconds: int = 8) -> None:
    with wave.open(str(input_path), "rb") as reader:
        params = reader.getparams()
        frame_count = min(reader.getnframes(), int(reader.getframerate() * seconds))
        frames = reader.readframes(frame_count)

    output_path.parent.mkdir(parents=True, exist_ok=True)
    with wave.open(str(output_path), "wb") as writer:
        writer.setparams(params)
        writer.writeframes(frames)


def main() -> None:
    profile = json.loads(PROFILE_PATH.read_text(encoding="utf-8-sig"))
    REFERENCE_DIR.mkdir(parents=True, exist_ok=True)

    for sample in profile.get("samples", []):
        source = sample.get("normalizedPath") or sample.get("path")
        if not source:
            continue

        output_path = REFERENCE_DIR / f"{sample['id']}-ref.wav"
        trim_wav(Path(source), output_path)
        sample["referencePath"] = str(output_path.resolve())
        sample["referencePromptText"] = REFERENCE_PROMPT

    PROFILE_PATH.write_text(json.dumps(profile, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")
    print(json.dumps(profile.get("samples", [])[-1], ensure_ascii=False, indent=2))


if __name__ == "__main__":
    main()

File: scripts/create-reference-samples.mjs
import fs from "node:fs";
import path from "node:path";

const profilePath = "auto_speech_data/profiles/default/profile.json";
const referenceDir = "auto_speech_data/profiles/default/reference";
const referencePromptText = "这是我的声音样本，用来帮助软件生成自然清晰的中文旁白。";

function readUInt32LE(buffer, offset) {
  return buffer.readUInt32LE(offset);
}

function trimWav(inputPath, outputPath, seconds = 8) {
  const buffer = fs.readFileSync(inputPath);
  if (buffer.toString("ascii", 0, 4) !== "RIFF" || buffer.toString("ascii", 8, 12) !== "WAVE") {
    throw new Error(`Not a WAV file: ${inputPath}`);
  }

  let offset = 12;
  let fmtChunk;
  let dataChunk;

  while (offset + 8 <= buffer.length) {
    const id = buffer.toString("ascii", offset, offset + 4);
    const size = readUInt32LE(buffer, offset + 4);
    const start = offset + 8;
    if (id === "fmt ") {
      fmtChunk = { start, size };
    }
    if (id === "data") {
      dataChunk = { start, size };
      break;
    }
    offset = start + size + (size % 2);
  }

  if (!fmtChunk || !dataChunk) {
    throw new Error(`Incomplete WAV file: ${inputPath}`);
  }

  const channels = buffer.readUInt16LE(fmtChunk.start + 2);
  const sampleRate = buffer.readUInt32LE(fmtChunk.start + 4);
  const bitsPerSample = buffer.readUInt16LE(fmtChunk.start + 14);
  const bytesPerFrame = channels * (bitsPerSample / 8);
  const keepBytes = Math.min(dataChunk.size, sampleRate * seconds * bytesPerFrame);

  const output = Buffer.concat([
    buffer.subarray(0, dataChunk.start),
    buffer.subarray(dataChunk.start, dataChunk.start + keepBytes),
  ]);

  output.writeUInt32LE(output.length - 8, 4);
  output.writeUInt32LE(keepBytes, dataChunk.start - 4);
  fs.mkdirSync(path.dirname(outputPath), { recursive: true });
  fs.writeFileSync(outputPath, output);
}

const profile = JSON.parse(fs.readFileSync(profilePath, "utf8").replace(/^\uFEFF/, ""));
fs.mkdirSync(referenceDir, { recursive: true });

for (const sample of profile.samples ?? []) {
  const source = sample.normalizedPath || sample.path;
  if (!source) continue;
  const outputPath = path.resolve(referenceDir, `${sample.id}-ref.wav`);
  trimWav(source, outputPath);
  sample.referencePath = outputPath;
  sample.referencePromptText = referencePromptText;
}

fs.writeFileSync(profilePath, `${JSON.stringify(profile, null, 2)}\n`, "utf8");
console.log(JSON.stringify(profile.samples.at(-1), null, 2));
